5 add support for speech to text using whisper

iKachaykin opened the pull request:
@Warefoxe could you please take a look?

Warefoxe commented:
@iKachaykin, it is looks good, but have some thioughts about this:
1. IMHO better will pass samples stream to `predict`(maybe need to rename this function, IMHO) function directly, moreover we alwais can make stream from file. 
2. result should not save to file. 
   - If use current implementation ( samples array to text ) then predict should return resulting String 
   - If pass stream to predict then better, IMHO, make some callback that will accept result or maybe string stream so we can `read` it.

iKachaykin commented:
@Warefoxe Thanks for your feedback!
1. It already takes samples as an input -> https://github.com/nchapman/nebula/blob/5-add-support-for-speech-to-text-using-whisper/src/backend/whisper.rs#L38, or do you mean something else? For now it is just a vector of floats, should I use some other more complex object maybe?
2. Good catch, thanks for this suggestion. Will include this change.

So to sum up:
- input: audio samples as a vector of floats
- output: string with the text said in the samples
Does it work for your that way? Or would you prefer something else?

Warefoxe commented:
@iKachaykin, It looks good.

iKachaykin commented:
@Warefoxe Great, thanks for your feedback!

iKachaykin commented:
@nchapman @Warefoxe This one should finally be completed :tada:

Latest updates:
- Refactored and restructured the code according to the latest suggestions by @Warefoxe , so that it takes samples as an input and returns `String` as an output.
- Succeeded to run whisper.cpp on a stream inside nebula.
- Updated [README.org](https://github.com/nchapman/nebula/blob/5-add-support-for-speech-to-text-using-whisper/examples/README.org#whisper_on_wav) to show how to use it.

A couple of **important notes**:
- For now as our library is feature based in main, I only succeeded to run whisper and llama together by separating them into separate features. Hence, to run llama models we will need to pass `--features llama` to basic examples. Pleas check here: [link](https://github.com/nchapman/nebula/blob/5-add-support-for-speech-to-text-using-whisper/examples/README.org#usage). @Warefoxe I am open to your suggestions as usual for another solution how to run them simultaneously without passing features.
- In the streaming mode, whisper will now dump a tmp wav to drive, then read it and convert to text. Even though it technically works, I believe it is not optimal way of achieving the streaming feature. @Warefoxe I would ask you to take a look at this one as well when you have time. I almost succeeded to do it [here](https://github.com/nchapman/nebula/blob/de77bd969ff27dd0fefe81cf697a966838358e0e/examples/whisper_stream.rs#L56) without a wav tmp file, but it failed because samples variable was moved and I couldn't reuse it later [here](https://github.com/nchapman/nebula/blob/de77bd969ff27dd0fefe81cf697a966838358e0e/examples/whisper_stream.rs#L85). Do you maybe have some suggestions?

Please let me know when I may merge this PR to bring whisper to main :)

Warefoxe commented:
@iKachaykin,
1 thing, can we somehow get result by parts during recognition, maybe add some callback which will call on every segment?

Warefoxe commented on src/backend/whisper.rs:
```diff
@@ -0,0 +1,66 @@
+        let mut out = String::new();
+        for i in 0..num_segments {
+            let segment = state
+                .full_get_segment_text(i)
```
Here I think should  be callback call

Warefoxe commented:
@iKachaykin, I think we need to make it here,  in any case for pulsar integration, we will use the Nebula Tauri plugin and we able to update it there.

iKachaykin commented:
@Warefoxe done, rebased and merged according to the way that we discussed today on the call (if someone else reads that: for now we don't integrate callbacks and will merge it as it is, we will add callbacks later).

Could you please take a look when you have time?
